File: src/JavaScript/Selection/quick_select.js
class QuickSelect {
    /**
     * Returns the kth smallest element in the array.
     */
    static selectKthSmallest(array, k) {
        const length = array.length;
        if (k < 1 || k > length)
            return undefined;
        return this.quickSelect(array, 0, length - 1, k, true);
    }

    /**
     * Returns the kth largest element in the array.
     */
    static selectKthLargest(array, k) {
        const length = array.length;
        if (k < 1 || k > length)
            return undefined;
        return this.quickSelect(array, 0, length - 1, k, false);
    }

    static quickSelect(array, left, right, k, isSmallest) {
        if (left === k)
            return array[left];
        const pivotIndex = this.partition(array, left, right, isSmallest);
        const order = pivotIndex - left + 1;
        if (k === order)
            return array[pivotIndex];
        else if (k < order)
            return this.quickSelect(array, left, pivotIndex - 1, k, isSmallest);
        else
            return this.quickSelect(array, pivotIndex + 1, right, k - order, isSmallest);
    }

    static partition(array, left, right, isSmallest) {
        const pivot = array[right];
        let i = left - 1;
        for (let j = left; j < right; j++) {
            if ((isSmallest && array[j] <= pivot) || (!isSmallest && array[j] >= pivot)) {
                i++;
                this.swap(array, i, j);
            }
        }
        this.swap(array, i + 1, right);
        return i + 1;
    }

    static swap(array, i, j) {
        [array[i], array[j]] = [array[j], array[i]]
    }
}

module.exports = QuickSelect

File: src/Python/Selection/quick_select.py
class QuickSelect:

    @staticmethod
    def select_kth_smallest(array, k):
        """ Returns the kth smallest element in the array. """
        if k < 1 or len(array) < k:
            return None
        return QuickSelect.quick_select(array, 0, len(array) - 1, k , True)

    @staticmethod
    def select_kth_largest(array, k):
        """ Returns the kth largest element in the array. """
        if k < 1 or len(array) < k:
            return None
        return QuickSelect.quick_select(array, 0, len(array) - 1, k, False)
    
    @staticmethod
    def quick_select(array, left, right, k, is_smallest):
        if left == right:
            return array[left]
        pivot_index = QuickSelect.partition(array, left, right, is_smallest)
        order = pivot_index - left + 1
        if k == order:
            return array[pivot_index]
        if k < order:
            return QuickSelect.quick_select(array, left, pivot_index - 1, k, is_smallest)
        return QuickSelect.quick_select(array, pivot_index + 1, right, k - order, is_smallest)
    
    @staticmethod
    def partition(array, left, right, is_smallest):
        pivot = array[right]
        i = left - 1
        for j in range(left, right):
            if (is_smallest and array[j] <= pivot) or (not is_smallest and array[j] >= pivot):
                i += 1
                QuickSelect.swap(array, i, j)
        QuickSelect.swap(array, i + 1, right)
        return i + 1

    @staticmethod
    def swap(array, i, j):
        array[i], array[j] = array[j], array[i]